File: JavaScript/Medium/graphs/numWaysArriveDestination.js
var solution = function(n, roads){
    let graph = {}
    let mod = 1000000007
    let start = 0
    
    for (const [n1, n2, weight] of roads){
        if (!graph[n1]) graph[n1] = {}
        if (!graph[n2]) graph[n2] = {}
        graph[n1][n2] = weight
        graph[n2][n1] = weight
    }

    let pq = new MinPriorityQueue({ priority: x => x[0]})
    pq.enqueue([0, start])

    let minCosts = new Array(n).fill(Infinity)
    let ways = new Array(n).fill(0)
    ways[start] = 1
    minCosts[start] = 0

    while (!pq.isEmpty()){
        let [cost, cur] = pg.dequeue().element;
        for (const [neighbor, dist] of Object.entries(graph[node])){
            let newPath = dist + cost
            if (newPath < minCosts[neighbor]){
                minCosts[neighbor] = newPath
                ways[neighbor] = ways[cur]
                pq.enqueue([newPath, neighbor])
            } else if (newPath === minCosts[neighbor]){
                ways[neighbor] += ways[cur]
                ways[neighbor] %= mod
            }
        }
    }

    return ways[n-1]
}

File: Python/Medium/graphs/numWaysArriveDestination.py
import collections
import heapq

def solution(n, roads):
    mod = 1000000007
    graph = collections.defaultdict(dict)
    start = 0

    for n1,n2,weight in roads:
        graph[n1][n2] = weight
        graph[n2][n1] = weight

    # (cost, startNode)
    pq = [(0, start)]
    minCosts = [float('inf')]*n
    minCosts[start] = 0
    ways = [0]*n
    ways[start]=1

    while pq:
        cost, cur = heapq.heappop(pq)
        for neighbor, dist in graph[cur].items():
            newPath = cost + dist
            if newPath < minCosts[neighbor]:
                minCosts[neighbor] = newPath
                ways[neighbor] = ways[cur]
                heapq.heappush(pq, (newPath, neighbor))
            elif newPath == minCosts[neighbor]:
                ways[neighbor] += ways[cur]
                ways[neighbor] %= mod

    return ways[n-1]

